### PR TITLE
new module plink genome

### PR DESCRIPTION
**Pull request by RuthEberhardt**
Apologies in my organisation I cannot test with conda or docker - the module works with singularity and I have loaded plink in the same was as existing modules so don't anticipate a problem

## PR checklist

Closes #XXX 

- [y ] This comment contains a description of changes (with reason).
- [y ] If you've fixed a bug or added code that should be tested, add tests!
- [x] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [x] If necessary, include test data in your PR.
- [y ] Remove all TODO statements.
- [y ] Emit the `versions.yml` file.
- [y] Follow the naming conventions.
- [y ] Follow the parameters requirements.
- [y ] Follow the input/output options guidelines.
- [x] Add a resource `label`
- [x] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
  - For modules:
    - [ ] `nf-core modules test <MODULE> --profile docker`
    - [ ] `nf-core modules test <MODULE> --profile singularity`
    - [ ] `nf-core modules test <MODULE> --profile conda`
  - For subworkflows:
    - [ ] `nf-core subworkflows test <SUBWORKFLOW> --profile docker`
    - [ ] `nf-core subworkflows test <SUBWORKFLOW> --profile singularity`
    - [ ] `nf-core subworkflows test <SUBWORKFLOW> --profile conda`

### CODE REVIEW

File: modules/nf-core/plink/genome/main.nf [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+    output:
+    tuple val(meta), path("*.genome")   , emit: genome
+    path "versions.yml"                 , emit: versions
+
+    when:
+    task.ext.when == null || task.ext.when
+
+    script:
+    def args = task.ext.args ?: ''
```
Comment (itrujnara) [inline modules/nf-core/plink/genome/main.nf:21]: Align equal signs

File: modules/nf-core/plink/genome/main.nf [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    conda "${moduleDir}/environment.yml"
+    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
+        'https://depot.galaxyproject.org/singularity/plink:1.90b6.21--h031d066_5':
+        'biocontainers/plink:1.90b6.21--h031d066_5' }"
+
+    input:
+    tuple val(meta), path(bed), path(bim), path(fam)
+
+    output:
+    tuple val(meta), path("*.genome")   , emit: genome
```
Comment (itrujnara) [inline modules/nf-core/plink/genome/main.nf:14]: Move the emits closer, the comma should be right after the longest text

**Review by itrujnara — APPROVED** [review]
Some formatting issues but good to go

File: modules/nf-core/plink/genome/main.nf [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    task.ext.when == null || task.ext.when
+
+    script:
+    def args = task.ext.args ?: ''
+    def prefix = task.ext.prefix ?: "${meta.id}"
+    if( "$bed" == "${prefix}.bed" ) error "Input and output names are the same, use \"task.ext.prefix\" in modules.config to disambiguate!"
+
+    """
+    plink \\
+        --bfile ${bed.baseName} \\
```
Comment (Joon-Klaps) [inline modules/nf-core/plink/genome/main.nf:27]:
```suggestion
        --bed ${bed} \\
        --bim ${bim} \\
        --fam ${fam} \\
```
Better to do it explicitly. Names could be inconsistent across the files.

File: modules/nf-core/plink/genome/main.nf [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    tag "$meta.id"
+    label 'process_medium'
+
+    conda "${moduleDir}/environment.yml"
+    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
+        'https://depot.galaxyproject.org/singularity/plink:1.90b6.21--h031d066_5':
+        'biocontainers/plink:1.90b6.21--h031d066_5' }"
+
+    input:
+    tuple val(meta), path(bed), path(bim), path(fam)
```
Comment (Joon-Klaps) [inline modules/nf-core/plink/genome/main.nf:11]: I'm not as familiar with the tool, could you also just provide a `vcf`,`bcf` or `phe` file instead? 
If so, consider adding them as input variables. For example: 
https://github.com/nf-core/modules/blob/master/modules/nf-core/plink/epistasis/main.nf

If you can - add tests for those input files.

File: modules/nf-core/plink/genome/main.nf [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    tuple val(meta), path("*.genome")   , emit: genome
+    path "versions.yml"                 , emit: versions
+
+    when:
+    task.ext.when == null || task.ext.when
+
+    script:
+    def args = task.ext.args ?: ''
+    def prefix = task.ext.prefix ?: "${meta.id}"
+    if( "$bed" == "${prefix}.bed" ) error "Input and output names are the same, use \"task.ext.prefix\" in modules.config to disambiguate!"
```
Comment (Joon-Klaps) [inline modules/nf-core/plink/genome/main.nf:23]:
```suggestion
```
This is likely redundant no - but only if no bed files are being created.

File: modules/nf-core/plink/genome/meta.yml [hunk 6]
```diff
@@ -0,0 +1,53 @@
+name: "plink_genome"
+description: Calculates identity-by-descent
+keywords:
+  - plink
+  - identity-by-descent
+  - genetics
```
Comment (Joon-Klaps) [inline modules/nf-core/plink/genome/meta.yml:6]:
```suggestion
  - genetics
  - genome
```

File: modules/nf-core/plink/genome/meta.yml [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+        description: PLINK binary genotype table file
+        pattern: "*.{bed}"
+    - bim:
+        type: file
+        description: PLINK extended MAP file
+        pattern: "*.{bim}"
+    - fam:
+        type: file
+        description: PLINK sample information file
+        pattern: "*.{fam}"
```
Comment (Joon-Klaps) [inline modules/nf-core/plink/genome/meta.yml:33]: If more inputs can be given, don't forget to add them here

File: modules/nf-core/plink/genome/main.nf [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
+        'https://depot.galaxyproject.org/singularity/plink:1.90b6.21--h031d066_5':
+        'biocontainers/plink:1.90b6.21--h031d066_5' }"
+
+    input:
+    tuple val(meta), path(bed), path(bim), path(fam)
+
+    output:
+    tuple val(meta), path("*.genome")   , emit: genome
+    path "versions.yml"                 , emit: versions
```
Comment (Joon-Klaps) [inline modules/nf-core/plink/genome/main.nf]: no need to have those additional spaces.

File: modules/nf-core/plink/genome/tests/main.nf.test [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+    tag "modules_nfcore"
+    tag "plink"
+    tag "plink/genome"
+
+    
+    test("plink - genome") {
+        when {
+            params {
+                outdir = "test"
+            }
```
Comment (Joon-Klaps) [inline modules/nf-core/plink/genome/tests/main.nf.test:17]:
```suggestion
```
No need for this 😄, `params.outdir` is mostly used on a pipeline level.

Comment (sateeshperi) [inline modules/nf-core/plink/genome/tests/main.nf.test:17]:
```suggestion
```

**Review by sateeshperi — CHANGES_REQUESTED** [review]
plz add a stub and a stub-test

File: modules/nf-core/plink/genome/meta.yml [hunk 10]
```diff
@@ -0,0 +1,53 @@
+name: "plink_genome"
+description: Calculates identity-by-descent
```
Comment (Joon-Klaps) [inline modules/nf-core/plink/genome/meta.yml:2]:
```suggestion
description: Calculates identity-by-descent over autosomal SNPs
```

File: modules/nf-core/plink/genome/main.nf (hunk 1 above)
Comment (RuthEberhardt) [reply modules/nf-core/plink/genome/main.nf]: Thank you, Ive fixed this

File: modules/nf-core/plink/genome/main.nf (hunk 2 above)
Comment (RuthEberhardt) [reply modules/nf-core/plink/genome/main.nf]: Thank you, I've fixed this

**Review by Joon-Klaps — CHANGES_REQUESTED** [review]
Very nice contribution! 
Minor suggestions 👍

File: modules/nf-core/plink/genome/tests/main.nf.test (hunk 9 above)
Comment (RuthEberhardt) [reply modules/nf-core/plink/genome/tests/main.nf.test:17]: Thank you, I will have to find documentation for this as I am new to all this.

File: modules/nf-core/plink/genome/main.nf (hunk 3 above)
Comment (RuthEberhardt) [reply modules/nf-core/plink/genome/main.nf:27]: Thank you, I've fixed this

File: modules/nf-core/plink/genome/main.nf (hunk 4 above)
Comment (RuthEberhardt) [reply modules/nf-core/plink/genome/main.nf:11]: Thank you, this particular plink function requires bed, bim and fam and doesn't use bcf/vcf/phe

File: modules/nf-core/plink/genome/tests/main.nf.test (hunk 9 above)
Comment (RuthEberhardt) [reply modules/nf-core/plink/genome/tests/main.nf.test:17]: Thank you, I've fixed this

File: modules/nf-core/plink/genome/meta.yml (hunk 10 above)
Comment (RuthEberhardt) [reply modules/nf-core/plink/genome/meta.yml:2]: Thank you!

File: modules/nf-core/plink/genome/main.nf (hunk 8 above)
Comment (RuthEberhardt) [reply modules/nf-core/plink/genome/main.nf]: Thank you, I've incorporated your suggestions

File: modules/nf-core/plink/genome/main.nf [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+        --out ${prefix}
+
+    cat <<-END_VERSIONS > versions.yml
+    "${task.process}":
+        plink: \$(echo \$(plink --version 2>&1) | sed 's/^PLINK v//' | sed 's/..-bit.*//' )
+    END_VERSIONS
+    """
+
+    stub:
+    def args   = task.ext.args ?: ''
```
Comment (Joon-Klaps) [inline modules/nf-core/plink/genome/main.nf:40]:
```suggestion
```
The language server will likely mention that these aren't being used. So think you can remove.

File: modules/nf-core/plink/genome/main.nf [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+    script:
+    def args   = task.ext.args ?: ''
+    def prefix = task.ext.prefix ?: "${meta.id}"
+    """
+    plink \\
+        --bed ${bed} \\
+        --bim ${bim} \\
+        --fam ${fam} \\
+        $args \\
+        --genome \\
```
Comment (Joon-Klaps) [inline modules/nf-core/plink/genome/main.nf:29]:
```suggestion
         --genome \\
         --bed ${bed} \\
        --bim ${bim} \\
        --fam ${fam} \\
        $args \\
```

**Review by Joon-Klaps — CHANGES_REQUESTED** [review]
Update the snapshot 😉 to include the stub.

File: modules/nf-core/plink/genome/main.nf (hunk 11 above)
Comment (sateeshperi) [reply modules/nf-core/plink/genome/main.nf:40]: $args are use in the command

Comment (RuthEberhardt) [reply modules/nf-core/plink/genome/main.nf:40]: Does it being used in the command mean it is needed for the stub? This is all new to me.

Comment (sateeshperi) [reply modules/nf-core/plink/genome/main.nf:40]: oh sorry this is the stub...so its not being used.... agree with @Joon-Klaps can be removed

**Comment by sateeshperi** [comment]
@RuthEberhardt plz update snapshot after adding stub test

**Review by sateeshperi — APPROVED** [review]

**Pull request merged**